motionplan/collision.go

```diff
@@ -49,301 +46,103 @@ func collisionListsAlmostEqual(cs1, cs2 []Collision) bool {
-	if !encompassed {
-		return 1, nil
+func newGeometryGraph(x, y map[string]spatial.Geometry, reportDistances bool) geometryGraph {
+	distances := make(map[string]map[string]float64)
+	for name := range x {
+		distances[name] = make(map[string]float64)
 	}
-	return -1, nil
-}
-
-func (sce spaceCollisionEntities) reportCollisions(distances []float64) []int {
-	collisionIndices := make([]int, 0)
-	for i := range distances {
-		if distances[i] >= spatial.CollisionBuffer {
-			collisionIndices = append(collisionIndices, i)
-		} else {
-			return []int{}
-		}
+	return geometryGraph{
+		x:               x,
+		y:               y,
+		distances:       distances,
+		reportDistances: reportDistances,
 	}
-	return collisionIndices
 }
 
-// collisionGraph is an implementation of an undirected graph used to track collisions between two set of CollisionEntities.
 type collisionGraph struct {
-	// key CollisionEntities
-	key *ObjectCollisionEntities
-
-	// test CollisionEntities are the set of CollisionEntities from which the collisionGraph takes its
-	// collisionCheckFn and collisionReportFn functions to check and report collisions between the
-	// test CollisionEntities and key Collision Entities
-	test CollisionEntities
-
-	// adjacencies is 2D array encoding edges between collisionEntiies in the collisionGraph.
-	// if adjacencies[i][j] >= 0 this corresponds to an edge between the entities at key[i] and test[j]
-	adjacencies [][]float64
-
-	// triangular is a bool that describes if the adjacencies matrix is triangular, which will be the case when key == test
-	triangular bool
-
-	reportDistances bool
+	geometryGraph
 }
 
 // newCollisionGraph instantiates a collisionGraph object and checks for collisions between the key and test sets of CollisionEntities
 // collisions that are reported in the reference CollisionSystem argument will be ignore and not stored as edges in the graph.
-func newCollisionGraph(
-	key *ObjectCollisionEntities,
-	test CollisionEntities,
-	reference *CollisionSystem,
-	reportDistances bool,
-) (*collisionGraph, error) {
-	var err error
-	cg := &collisionGraph{
-		key:             key,
-		test:            test,
-		adjacencies:     make([][]float64, key.count()),
-		triangular:      key == test,
-		reportDistances: reportDistances,
+func newCollisionGraph(x, y map[string]spatial.Geometry, reference *collisionGraph, reportDistances bool) (cg *collisionGraph, err error) {
+	if y == nil {
+		y = x
 	}
-	for i := range cg.adjacencies {
-		cg.adjacencies[i] = make([]float64, test.count())
-		keyi := key.entityFromIndex(i)
-		startIndex := 0
-		if cg.triangular {
-			startIndex = i + 1
-			for j := 0; j < startIndex; j++ {
-				cg.adjacencies[i][j] = math.NaN()
+	cg = &collisionGraph{newGeometryGraph(x, y, reportDistances)}
+
+	var distance float64
+	for xName, xGeometry := range cg.x {
+		for yName, yGeometry := range cg.y {
+			if _, ok := cg.distances[yName][xName]; ok || xGeometry == yGeometry {
```

Comment:
Can we also make a setter/getter for distances?

That way all ordering is handled behind the scenes inside the cg, in one single place.